### PR TITLE
Website guides on HD key and address generation

### PR DESCRIPTION
**Pull request by torkelrogstad**
I also slightly clean up TransactionTest, as I stumbled on that while writing the guides.

### CODE REVIEW

File: docs/core/addresses.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+val segwitAddress = {
+    // see https://bitcoin.org/en/glossary/pubkey-script
+    // for reading resources on the details of scriptPubKeys
+    // pay-to-witness-pubkey-hash scriptPubKey V0
+    val scriptPubKey = P2WPKHWitnessSPKV0(pubkey)
+    Bech32Address(scriptPubKey, TestNet3)
+}
+```
+
+## Generating legacy (base58) addresses
```
Comment (nkohen) [inline docs/core/addresses.md:49]: Just because it's so much simpler I'd argue this section should go above segwit

File: docs/core/addresses.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+val privkey = ECPrivateKey()
+val pubkey = privkey.publicKey
+
+val segwitAddress = {
+    // see https://bitcoin.org/en/glossary/pubkey-script
+    // for reading resources on the details of scriptPubKeys
+    // pay-to-witness-pubkey-hash scriptPubKey V0
+    val scriptPubKey = P2WPKHWitnessSPKV0(pubkey)
+    Bech32Address(scriptPubKey, TestNet3)
+}
```
Comment (nkohen) [inline docs/core/addresses.md:46]: Any reason not to have this be
```
// see https://bitcoin.org/en/glossary/pubkey-script
// for reading resources on the details of scriptPubKeys
// pay-to-witness-pubkey-hash scriptPubKey V0
val scriptPubKey = P2WPKHWitnessSPKV0(pubkey)

val segwitAddress = Bech32Address(scriptPubKey, TestNet3)
```

File: docs/core/hd-keys.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+id: hd-keys
+title: HD key generation
+---
+
+In modern Bitcoin wallets, users only need to write down
+a sequence of words, and that sequence is a complete backup
+of their wallet. This is thanks to what's called Hierarchical
+Deterministic key generation. In short, every wallet using HD
+key generation has a root seed for each wallet, and this
+seed can be used to generate an infinite amount of later
```
Comment (nkohen) [inline docs/core/hd-keys.md]: nit: replace infinite with arbitrary ;)

File: docs/core/hd-keys.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+> If you want to jump into the details of how this work,
+> you should check out
+> [BIP 32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki).
+
+Bitcoin-S supports generating keys in this fashion. Here's a
+full example of how to obtain a wallet seed, and then
+use that to generate further private and public keys:
+
+Here's an example:
```
Comment (nkohen) [inline docs/core/hd-keys.md]: You say here's an example twice (once in each paragraph)

File: docs/core/hd-keys.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+val segwitPath = SegWitHDPath.fromString("m/84'/0'/0'/0/0")
+
+// alternatively:
+val otherSegwitPath =
+SegWitHDPath(HDCoinType.Bitcoin,
+             accountIndex = 0,
+             HDChainType.External,
+             addressIndex = 0)
+
+segwitPath == otherSegwitPath
```
Comment (nkohen) [inline docs/core/hd-keys.md:63]: either wrap in `assert` or comment `// true` for clarity

File: docs/rpc/rpc-clients-intro.md [hunk 6]
```diff
@@ -3,4 +3,27 @@ id: rpc-clients-intro
...
+have a custom type that represents a Bitcoin transaction,
+how do you get that to play nicely with the result that
+Bitcoin Core gives you after signing a transaction? A
+random hexadecimal string in a HTTP response could be
+anything from a public key, a transaction or a block
+header.
+
+We've done all the mundane work of wiring requests and
+responses from Bitcoin Core to the powerful and safe types
+found in Bitcoin-S. You'll know for sure that you're sending
```
Comment (nkohen) [inline docs/rpc/rpc-clients-intro.md]: mention tests too?

File: docs/core/addresses.md (hunk 2 above)
Comment (torkelrogstad) [reply docs/core/addresses.md:46]: I was thinking I'd have a variable named `scriptPubKey` beneath here as well, but that didn't end up happening. do you think it's confusing with a block with vals inside them?

File: docs/core/addresses.md (hunk 1 above)
Comment (torkelrogstad) [reply docs/core/addresses.md:49]: We want to push people towards segwit though! And That's also one of the differentiating things with Bitcoin-S, how easy segwit addresses are. Notice that the code in the block below reuses the values from the block above, so the only lines that are different are 
```scala
val scriptPubKey = P2WPKHWitnessSPKV0(pubkey)
    Bech32Address(scriptPubKey, TestNet3)
```
and 
```scala
 val legacyAddress = P2PKHAddress(pubkey, TestNet3)
```

File: docs/core/hd-keys.md (hunk 5 above)
Comment (torkelrogstad) [reply docs/core/hd-keys.md:63]: So this is one of the cool things about mdoc (https://github.com/scalameta/mdoc), it typechecks and even runs the code before publishing. So `// true` gets added at the bottom by the compiler

File: docs/rpc/rpc-clients-intro.md (hunk 6 above)
Comment (torkelrogstad) [reply docs/rpc/rpc-clients-intro.md]: good point

File: docs/core/addresses.md (hunk 1 above)
Comment (nkohen) [reply docs/core/addresses.md:49]: Ah, good points :)

File: docs/core/addresses.md (hunk 2 above)
Comment (nkohen) [reply docs/core/addresses.md:46]: I think it's marginally more work looking back and forth to see what is what, not a big deal obviously

File: docs/core/hd-keys.md (hunk 5 above)
Comment (nkohen) [reply docs/core/hd-keys.md:63]: gotchya, very cool!

**Pull request merged**